File: src/tools/shared.ts
import type { WorkflowSummary } from "../types.js";

/**
 * Convert a Temporal workflow listing entry to a WorkflowSummary.
 */
export function toWorkflowSummary(workflow: {
  workflowId: string;
  runId: string;
  type: string;
  status: { name: string };
  startTime?: Date | null;
  closeTime?: Date | null;
  taskQueue: string;
  searchAttributes?: Record<string, unknown>;
}): WorkflowSummary {
  const startTime = workflow.startTime?.toISOString() ?? "";
  const closeTime = workflow.closeTime?.toISOString() ?? null;
  const durationMs =
    workflow.startTime && workflow.closeTime
      ? workflow.closeTime.getTime() - workflow.startTime.getTime()
      : null;

  return {
    workflow_id: workflow.workflowId,
    run_id: workflow.runId,
    workflow_type: workflow.type,
    status: workflow.status.name,
    start_time: startTime,
    close_time: closeTime,
    duration_ms: durationMs,
    task_queue: workflow.taskQueue,
    search_attributes: { ...workflow.searchAttributes },
  };
}

File: src/temporal_xray/tools/shared.py
from __future__ import annotations

from typing import Any

from ..types import WorkflowSummary


def to_workflow_summary(workflow: Any) -> WorkflowSummary:
    """Convert a Temporal workflow listing entry to a WorkflowSummary."""
    start_time = workflow.start_time.isoformat() if workflow.start_time else ""
    close_time = workflow.close_time.isoformat() if workflow.close_time else None
    duration_ms = None
    if workflow.start_time and workflow.close_time:
        delta = workflow.close_time - workflow.start_time
        duration_ms = int(delta.total_seconds() * 1000)

    search_attrs = {}
    if workflow.search_attributes:
        # temporalio Python SDK returns SearchAttributes as a mapping
        for key, values in workflow.search_attributes.items():
            search_attrs[key] = values[0] if len(values) == 1 else list(values)

    return WorkflowSummary(
        workflow_id=workflow.id,
        run_id=workflow.run_id,
        workflow_type=workflow.workflow_type,
        status=workflow.status.name if hasattr(workflow.status, "name") else str(workflow.status),
        start_time=start_time,
        close_time=close_time,
        duration_ms=duration_ms,
        task_queue=workflow.task_queue,
        search_attributes=search_attrs,
    )
